### PR TITLE
Add BusinessId api response model

### PR DESCRIPTION
**Pull request by vinaysshenoy**
Pivotal Story Link: https://www.pivotaltracker.com/story/show/164266803

### CODE REVIEW

**Comment by saket** [comment]
@vinaysshenoy does this feature have an associated pivotal story?

**Comment by vinaysshenoy** [comment]
> @vinaysshenoy does this feature have an associated pivotal story?

Whoops, I keep forgetting to do this. Thanks for reminding me!

File: app/src/main/java/org/simple/clinic/patient/BusinessId.kt [hunk 1]
```diff
@@ -35,39 +38,70 @@ data class BusinessId(
...
 
     class RoomTypeConverter {
 
       @TypeConverter
-      fun fromStringValue(stringValue: String) = try {
-        MetaVersion.valueOf(stringValue)
-      } catch (e: IllegalArgumentException) {
-        UNKNOWN
-      }
+      fun toEnum(enum: String?) = TypeAdapter.toEnum(enum)
```
Comment (sanchita) [inline app/src/main/java/org/simple/clinic/patient/BusinessId.kt]: Same comment as above. Rename `enum` -> `stringValue`

File: app/src/main/java/org/simple/clinic/patient/BusinessId.kt [hunk 2]
```diff
@@ -35,39 +38,70 @@ data class BusinessId(
...
 
       @TypeConverter
-      fun toStringValue(identifierType: IdentifierType) = identifierType.name
+      fun fromEnum(enum: IdentifierType?) = TypeAdapter.fromEnum(enum)
+    }
+
+    class MoshiTypeAdapter {
+
+      @FromJson
+      fun toEnum(enum: String?) = TypeAdapter.toEnum(enum)
```
Comment (sanchita) [inline app/src/main/java/org/simple/clinic/patient/BusinessId.kt]: Same comment as above. Rename `enum` -> `stringValue`

File: app/src/main/java/org/simple/clinic/overdue/AppointmentCancelReason.kt [hunk 3]
```diff
@@ -43,42 +44,25 @@ sealed class AppointmentCancelReason {
...
+      unknownEnumToStringConverter = { (it as Unknown).actualValue }
+  )
 
-    fun fromEnum(reason: AppointmentCancelReason?): String? {
-      if (reason == null) {
-        return null
-      }
+  class RoomTypeConverter {
+    @TypeConverter
+    fun toEnum(reason: String?): AppointmentCancelReason? = TypeAdapter.toEnum(reason)
```
Comment (sanchita) [inline app/src/main/java/org/simple/clinic/overdue/AppointmentCancelReason.kt]: Rename `reason` -> `reasonString`

File: app/src/main/java/org/simple/clinic/util/SafeEnumTypeConverter.kt [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * This class is meant to provide a safe way to (de)serialise enum values from network responses
+ * without breaking older app versions that may not have support for a newly added enum value.
+ **/
+open class SafeEnumTypeAdapter<T>(
+    val knownMappings: Map<T, String>,
+    private val unknownEnumToStringConverter: (T) -> String,
+    private val unknownStringToEnumConverter: (String) -> T
+) {
+
+  fun toEnum(enum: String?): T? {
```
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/util/SafeEnumTypeConverter.kt]: Done.

File: app/src/main/java/org/simple/clinic/overdue/AppointmentCancelReason.kt (hunk 3 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/overdue/AppointmentCancelReason.kt]: Renamed to `value`.

File: app/src/main/java/org/simple/clinic/patient/BusinessId.kt (hunk 2 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/patient/BusinessId.kt]: Renamed to `value`.

File: app/src/main/java/org/simple/clinic/patient/BusinessId.kt (hunk 1 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/patient/BusinessId.kt]: Renamed to `value`.

**Pull request merged**